File: storage.py
"""SQLite persistence + schema migrations for the Custodian engine.

Design choices:
  - The database mirrors the in-memory Store exactly. `save` does a full
    transactional rewrite of the data tables. For a personal backlog (tens to a
    few hundred items) this is microseconds and is the least bug-prone option:
    the file is always an exact image of the engine after each operation.
  - Schema evolution is a migration CHAIN. Version N means "N migrations
    applied." A fresh database starts at 0 and runs every migration up to
    SCHEMA_VERSION, so the create path and the upgrade path are the same code,
    and the migration chain is exercised on every fresh install.
  - This is what lets a newer build open an older file without data loss: it
    reads the stored version and runs only the migrations the file is missing.

Migration history:
  v1  baseline schema (the preview's model, persisted)
  v2  add `protected` to containers, and guarantee a protected "Misc" exists
      (promote an existing one named "Misc", else create it). This is the
      deliberate addition on top of the preview; it is a migration so that any
      pre-protected file upgrades cleanly, and so the first-run empty state
      (exactly one protected Misc) falls out of the chain.

The display version string (e.g. "0.01.00a") is a separate, human-facing thing
owned by the app UI; this integer schema_version is the only thing that drives
migrations.
"""

import json
import sqlite3
from engine import Store, Container, Task, Routine, DayItem


# ---------- meta / versioning ----------

def _ensure_meta(conn):
    conn.execute("CREATE TABLE IF NOT EXISTS meta (key TEXT PRIMARY KEY, value TEXT)")


def _get_version(conn) -> int:
    _ensure_meta(conn)
    row = conn.execute("SELECT value FROM meta WHERE key='schema_version'").fetchone()
    return int(row[0]) if row else 0


def _set_version(conn, v: int):
    conn.execute(
        "INSERT INTO meta(key,value) VALUES('schema_version',?) "
        "ON CONFLICT(key) DO UPDATE SET value=excluded.value",
        (str(v),),
    )


# ---------- migrations ----------

def _migration_to_v1(conn):
    conn.execute("""
        CREATE TABLE containers (
            id          INTEGER PRIMARY KEY,
            name        TEXT    NOT NULL,
            status      TEXT    NOT NULL,
            description TEXT    NOT NULL DEFAULT '',
            ord         INTEGER NOT NULL
        )""")
    conn.execute("""
        CREATE TABLE tasks (
            id           INTEGER PRIMARY KEY,
            name         TEXT    NOT NULL,
            container_id INTEGER NOT NULL,
            col          TEXT    NOT NULL,
            position     INTEGER NOT NULL,
            completed    INTEGER NOT NULL DEFAULT 0
        )""")
    conn.execute("""
        CREATE TABLE routines (
            id             INTEGER PRIMARY KEY,
            name           TEXT    NOT NULL,
            container_id   INTEGER NOT NULL,
            days_of_week   TEXT    NOT NULL DEFAULT '[]',
            time           TEXT,
            duration_hours REAL,
            active         INTEGER NOT NULL DEFAULT 1,
            archived       INTEGER NOT NULL DEFAULT 0,
            ord            INTEGER NOT NULL
        )""")
    conn.execute("""
        CREATE TABLE day_log (
            id          INTEGER PRIMARY KEY,
            date_key    TEXT    NOT NULL,
            source_type TEXT    NOT NULL,
            source_id   INTEGER NOT NULL,
            done        INTEGER NOT NULL DEFAULT 0,
            one_off     INTEGER NOT NULL DEFAULT 0,
            seq         INTEGER NOT NULL
        )""")
    conn.execute("CREATE INDEX idx_daylog_date ON day_log(date_key, seq)")


def _migration_to_v2(conn):
    conn.execute("ALTER TABLE containers ADD COLUMN protected INTEGER NOT NULL DEFAULT 0")
    # Guarantee a protected Misc so the project always keeps a home for tasks.
    has_protected = conn.execute("SELECT 1 FROM containers WHERE protected=1 LIMIT 1").fetchone()
    if not has_protected:
        existing_misc = conn.execute(
            "SELECT id FROM containers WHERE name='Misc' ORDER BY id LIMIT 1"
        ).fetchone()
        if existing_misc:
            conn.execute("UPDATE containers SET protected=1 WHERE id=?", (existing_misc[0],))
        else:
            row = conn.execute("SELECT COALESCE(MAX(ord),-1)+1 FROM containers").fetchone()
            next_ord = row[0]
            conn.execute(
                "INSERT INTO containers(name,status,description,ord,protected) "
                "VALUES('Misc','active','',?,1)",
                (next_ord,),
            )


# Index i applies the upgrade to version i+1.
_MIGRATIONS = [_migration_to_v1, _migration_to_v2]
SCHEMA_VERSION = len(_MIGRATIONS)


def _apply_migrations(conn, up_to: int = None):
    target = SCHEMA_VERSION if up_to is None else up_to
    current = _get_version(conn)
    for i in range(current, target):
        _MIGRATIONS[i](conn)
        _set_version(conn, i + 1)
    conn.commit()


# ---------- load ----------

def load(conn) -> Store:
    conn.row_factory = sqlite3.Row
    store = Store()

    for r in conn.execute("SELECT * FROM containers ORDER BY ord, id"):
        store.containers[r["id"]] = Container(
            id=r["id"], name=r["name"], status=r["status"],
            description=r["description"], protected=bool(r["protected"]),
        )
        store.container_order.append(r["id"])

    for r in conn.execute("SELECT * FROM tasks"):
        store.tasks[r["id"]] = Task(
            id=r["id"], name=r["name"], container_id=r["container_id"],
            column=r["col"], position=r["position"], completed=bool(r["completed"]),
        )

    for r in conn.execute("SELECT * FROM routines ORDER BY ord, id"):
        store.routines[r["id"]] = Routine(
            id=r["id"], name=r["name"], container_id=r["container_id"],
            days_of_week=json.loads(r["days_of_week"]), time=r["time"],
            duration_hours=r["duration_hours"], active=bool(r["active"]),
            archived=bool(r["archived"]),
        )
        store.routine_order.append(r["id"])

    for r in conn.execute("SELECT * FROM day_log ORDER BY date_key, seq, id"):
        store.day_log.setdefault(r["date_key"], []).append(DayItem(
            id=r["id"], source_type=r["source_type"], source_id=r["source_id"],
            done=bool(r["done"]), one_off=bool(r["one_off"]),
        ))

    # resume id generation above every persisted id
    max_id = 0
    for d in (store.containers, store.tasks, store.routines):
        if d:
            max_id = max(max_id, max(d.keys()))
    for items in store.day_log.values():
        for it in items:
            max_id = max(max_id, it.id)
    store._next_id = max_id + 1
    return store


# ---------- save (full transactional rewrite) ----------

def save(store: Store, conn):
    cur = conn.cursor()
    cur.execute("BEGIN")
    try:
        for tbl in ("containers", "tasks", "routines", "day_log"):
            cur.execute(f"DELETE FROM {tbl}")

        ordered_cids = store.container_order + [cid for cid in store.containers
                                                if cid not in store.container_order]
        for ordn, cid in enumerate(ordered_cids):
            c = store.containers[cid]
            cur.execute(
                "INSERT INTO containers(id,name,status,description,ord,protected) "
                "VALUES(?,?,?,?,?,?)",
                (c.id, c.name, c.status, c.description, ordn, int(c.protected)),
            )

        for t in store.tasks.values():
            cur.execute(
                "INSERT INTO tasks(id,name,container_id,col,position,completed) "
                "VALUES(?,?,?,?,?,?)",
                (t.id, t.name, t.container_id, t.column, t.position, int(t.completed)),
            )

        ordered_rids = store.routine_order + [rid for rid in store.routines
                                              if rid not in store.routine_order]
        for ordn, rid in enumerate(ordered_rids):
            r = store.routines[rid]
            cur.execute(
                "INSERT INTO routines(id,name,container_id,days_of_week,time,"
                "duration_hours,active,archived,ord) VALUES(?,?,?,?,?,?,?,?,?)",
                (r.id, r.name, r.container_id, json.dumps(r.days_of_week), r.time,
                 r.duration_hours, int(r.active), int(r.archived), ordn),
            )

        for date_key, items in store.day_log.items():
            for seq, it in enumerate(items):
                cur.execute(
                    "INSERT INTO day_log(id,date_key,source_type,source_id,done,one_off,seq) "
                    "VALUES(?,?,?,?,?,?,?)",
                    (it.id, date_key, it.source_type, it.source_id,
                     int(it.done), int(it.one_off), seq),
                )

        _set_version(conn, SCHEMA_VERSION)
        conn.commit()
    except Exception:
        conn.rollback()
        raise


# ---------- entry point ----------

def open_database(path: str):
    """Open (or create) a database at `path`, run any pending migrations, and
    return (conn, store). A brand-new file ends up with exactly one protected
    Misc container, produced by the migration chain.

    check_same_thread=False: pywebview dispatches bridge calls on threads other
    than the one that opens the connection. The Api serializes every call with a
    lock, so the connection is only ever touched by one thread at a time."""
    conn = sqlite3.connect(path, check_same_thread=False)
    conn.row_factory = sqlite3.Row
    _apply_migrations(conn)
    store = load(conn)
    return conn, store

File: web/js/storage.js
// storage.js — SQLite persistence via sql.js + OPFS (port of storage.py)
// Exposes StewardStorage as a window global. No ES modules; loaded via <script> tag.

(function () {
  'use strict';

  const SCHEMA_VERSION = 2;

  // ---- meta / versioning ----

  function _ensureMeta(db) {
    db.run("CREATE TABLE IF NOT EXISTS meta (key TEXT PRIMARY KEY, value TEXT)");
  }

  function _getVersion(db) {
    _ensureMeta(db);
    const res = db.exec("SELECT value FROM meta WHERE key='schema_version'");
    if (!res.length || !res[0].values.length) return 0;
    return parseInt(res[0].values[0][0], 10);
  }

  function _setVersion(db, v) {
    db.run(
      "INSERT INTO meta(key,value) VALUES('schema_version',?) ON CONFLICT(key) DO UPDATE SET value=excluded.value",
      [String(v)]
    );
  }

  // ---- migrations ----

  function _migrationToV1(db) {
    db.run(`CREATE TABLE containers (
      id          INTEGER PRIMARY KEY,
      name        TEXT    NOT NULL,
      status      TEXT    NOT NULL,
      description TEXT    NOT NULL DEFAULT '',
      ord         INTEGER NOT NULL
    )`);
    db.run(`CREATE TABLE tasks (
      id           INTEGER PRIMARY KEY,
      name         TEXT    NOT NULL,
      container_id INTEGER NOT NULL,
      col          TEXT    NOT NULL,
      position     INTEGER NOT NULL,
      completed    INTEGER NOT NULL DEFAULT 0
    )`);
    db.run(`CREATE TABLE routines (
      id             INTEGER PRIMARY KEY,
      name           TEXT    NOT NULL,
      container_id   INTEGER NOT NULL,
      days_of_week   TEXT    NOT NULL DEFAULT '[]',
      time           TEXT,
      duration_hours REAL,
      active         INTEGER NOT NULL DEFAULT 1,
      archived       INTEGER NOT NULL DEFAULT 0,
      ord            INTEGER NOT NULL
    )`);
    db.run(`CREATE TABLE day_log (
      id          INTEGER PRIMARY KEY,
      date_key    TEXT    NOT NULL,
      source_type TEXT    NOT NULL,
      source_id   INTEGER NOT NULL,
      done        INTEGER NOT NULL DEFAULT 0,
      one_off     INTEGER NOT NULL DEFAULT 0,
      seq         INTEGER NOT NULL
    )`);
    db.run("CREATE INDEX idx_daylog_date ON day_log(date_key, seq)");
  }

  function _migrationToV2(db) {
    db.run("ALTER TABLE containers ADD COLUMN protected INTEGER NOT NULL DEFAULT 0");
    const hasProtected = db.exec("SELECT 1 FROM containers WHERE protected=1 LIMIT 1");
    if (!hasProtected.length || !hasProtected[0].values.length) {
      const existingMisc = db.exec("SELECT id FROM containers WHERE name='Misc' ORDER BY id LIMIT 1");
      if (existingMisc.length && existingMisc[0].values.length) {
        db.run("UPDATE containers SET protected=1 WHERE id=?", [existingMisc[0].values[0][0]]);
      } else {
        const ordRes = db.exec("SELECT COALESCE(MAX(ord),-1)+1 FROM containers");
        const nextOrd = ordRes.length ? ordRes[0].values[0][0] : 0;
        db.run(
          "INSERT INTO containers(name,status,description,ord,protected) VALUES('Misc','active','',?,1)",
          [nextOrd]
        );
      }
    }
  }

  const _MIGRATIONS = [_migrationToV1, _migrationToV2];

  function applyMigrations(db) {
    const current = _getVersion(db);
    for (let i = current; i < _MIGRATIONS.length; i++) {
      _MIGRATIONS[i](db);
      _setVersion(db, i + 1);
    }
  }

  // ---- load ----

  function load(db) {
    const { makeStore } = window.StewardEngine;
    const store = makeStore();

    const containers = db.exec("SELECT * FROM containers ORDER BY ord, id");
    if (containers.length) {
      const cols = containers[0].columns;
      for (const row of containers[0].values) {
        const r = Object.fromEntries(cols.map((c, i) => [c, row[i]]));
        store.containers[r.id] = { id: r.id, name: r.name, status: r.status, description: r.description, protected: !!r.protected };
        store.container_order.push(r.id);
      }
    }

    const tasks = db.exec("SELECT * FROM tasks");
    if (tasks.length) {
      const cols = tasks[0].columns;
      for (const row of tasks[0].values) {
        const r = Object.fromEntries(cols.map((c, i) => [c, row[i]]));
        store.tasks[r.id] = { id: r.id, name: r.name, container_id: r.container_id, column: r.col, position: r.position, completed: !!r.completed };
      }
    }

    const routines = db.exec("SELECT * FROM routines ORDER BY ord, id");
    if (routines.length) {
      const cols = routines[0].columns;
      for (const row of routines[0].values) {
        const r = Object.fromEntries(cols.map((c, i) => [c, row[i]]));
        store.routines[r.id] = {
          id: r.id, name: r.name, container_id: r.container_id,
          days_of_week: JSON.parse(r.days_of_week || "[]"),
          time: r.time || null, duration_hours: r.duration_hours != null ? r.duration_hours : null,
          active: !!r.active, archived: !!r.archived,
        };
        store.routine_order.push(r.id);
      }
    }

    const daylog = db.exec("SELECT * FROM day_log ORDER BY date_key, seq, id");
    if (daylog.length) {
      const cols = daylog[0].columns;
      for (const row of daylog[0].values) {
        const r = Object.fromEntries(cols.map((c, i) => [c, row[i]]));
        if (!store.day_log[r.date_key]) store.day_log[r.date_key] = [];
        store.day_log[r.date_key].push({ id: r.id, source_type: r.source_type, source_id: r.source_id, done: !!r.done, one_off: !!r.one_off });
      }
    }

    // reseed id generator above every persisted id
    let maxId = 0;
    for (const id of [...Object.keys(store.containers), ...Object.keys(store.tasks), ...Object.keys(store.routines)])
      if (Number(id) > maxId) maxId = Number(id);
    for (const items of Object.values(store.day_log))
      for (const it of items)
        if (it.id > maxId) maxId = it.id;
    store._next_id = maxId + 1;

    return store;
  }

  // ---- save (full transactional rewrite) ----

  function save(db, store) {
    db.run("BEGIN");
    try {
      for (const tbl of ["containers", "tasks", "routines", "day_log"]) {
        db.run(`DELETE FROM ${tbl}`);
      }

      const allCids = [
        ...store.container_order,
        ...Object.keys(store.containers).map(Number).filter(id => !store.container_order.includes(id)),
      ];
      allCids.forEach((cid, ordn) => {
        const c = store.containers[cid];
        if (!c) return;
        db.run(
          "INSERT INTO containers(id,name,status,description,ord,protected) VALUES(?,?,?,?,?,?)",
          [c.id, c.name, c.status, c.description, ordn, c.protected ? 1 : 0]
        );
      });

      for (const t of Object.values(store.tasks)) {
        db.run(
          "INSERT INTO tasks(id,name,container_id,col,position,completed) VALUES(?,?,?,?,?,?)",
          [t.id, t.name, t.container_id, t.column, t.position, t.completed ? 1 : 0]
        );
      }

      const allRids = [
        ...store.routine_order,
        ...Object.keys(store.routines).map(Number).filter(id => !store.routine_order.includes(id)),
      ];
      allRids.forEach((rid, ordn) => {
        const r = store.routines[rid];
        if (!r) return;
        db.run(
          "INSERT INTO routines(id,name,container_id,days_of_week,time,duration_hours,active,archived,ord) VALUES(?,?,?,?,?,?,?,?,?)",
          [r.id, r.name, r.container_id, JSON.stringify(r.days_of_week), r.time, r.duration_hours, r.active ? 1 : 0, r.archived ? 1 : 0, ordn]
        );
      });

      for (const [date_key, items] of Object.entries(store.day_log)) {
        items.forEach((it, seq) => {
          db.run(
            "INSERT INTO day_log(id,date_key,source_type,source_id,done,one_off,seq) VALUES(?,?,?,?,?,?,?)",
            [it.id, date_key, it.source_type, it.source_id, it.done ? 1 : 0, it.one_off ? 1 : 0, seq]
          );
        });
      }

      _setVersion(db, SCHEMA_VERSION);
      db.run("COMMIT");
    } catch (e) {
      db.run("ROLLBACK");
      throw e;
    }
  }

  // ---- OPFS persistence ----

  async function _readOpfs() {
    try {
      const root = await navigator.storage.getDirectory();
      const fh = await root.getFileHandle("steward.db", { create: false });
      const file = await fh.getFile();
      return new Uint8Array(await file.arrayBuffer());
    } catch {
      return null; // file doesn't exist yet
    }
  }

  async function persistDatabase(db) {
    const data = db.export();
    const root = await navigator.storage.getDirectory();
    const fh = await root.getFileHandle("steward.db", { create: true });
    const writable = await fh.createWritable();
    await writable.write(data);
    await writable.close();
  }

  async function openDatabase() {
    const bytes = await _readOpfs();
    const SQL = await initSqlJs({ locateFile: f => `./vendor/${f}` });
    const db = bytes ? new SQL.Database(bytes) : new SQL.Database();
    applyMigrations(db);
    const store = load(db);
    return { db, store };
  }

  // ---- export / import ----

  async function exportDatabase(db) {
    const data = db.export();
    const blob = new Blob([data], { type: "application/octet-stream" });
    const url = URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = "steward.db";
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    URL.revokeObjectURL(url);
  }

  async function importDatabase(file) {
    const data = new Uint8Array(await file.arrayBuffer());
    const SQL = await initSqlJs({ locateFile: f => `./vendor/${f}` });
    const db = new SQL.Database(data);
    applyMigrations(db); // handles v1 → v2 if importing an older desktop file
    const store = load(db);
    await persistDatabase(db);
    return { db, store };
  }

  window.StewardStorage = { openDatabase, persistDatabase, save, load, exportDatabase, importDatabase };

})();
